pkg/proxyagent/agent/agent.go

```diff
@@ -194,6 +199,67 @@ func GetClusterProxyValueFunc(
 		if err != nil {
 			return nil, err
 		}
+
+		// Get agentIndentifiers and servicesToExpose.
+		// agetnIdentifiers is used in `--agent-identifiers` flag in addon-agent-deployment.yaml.
+		// servicesToExpose defines the services we want to expose to the hub.
+
+		// List all available managedClusterSets
+		managedClusterSetList := &clusterv1beta1.ManagedClusterSetList{}
+		err = runtimeClient.List(context.TODO(), managedClusterSetList)
+		if err != nil {
+			return nil, err
+		}
+
+		// deleted managedClusterSets are not included in the list
+		managedClusterSets := filterManagedClusterSetByDeletionTimestamp(managedClusterSetList.Items)
+
+		// only cluseterSet cover current cluster include in the list.
+		managedClusterSets, err = filterManagedClusterSetForCurrentCluster(managedClusterSets, cluster.Labels)
+		if err != nil {
+			return nil, err
+		}
+
+		managedClusterSetMap := map[string]clusterv1beta1.ManagedClusterSet{}
+		for _, managedClusterSet := range managedClusterSets {
+			managedClusterSetMap[managedClusterSet.Name] = managedClusterSet
+		}
+
+		// List all available serviceResolvers
+		serviceResolverList := &proxyv1alpha1.ManagedProxyServiceResolverList{}
+		err = runtimeClient.List(context.TODO(), serviceResolverList)
+		if err != nil {
+			return nil, err
+		}
+
+		// only legal serviceResolvers include in the list.
+		serviceResolvers := filterLegalManagedProxyServiceResolver(serviceResolverList.Items)
+
+		// deleted serviceResolvers are not included in the list
+		serviceResolvers = filterManagedProxyServiceResolverByDeletionTimestamp(serviceResolvers)
+
+		// filter serviceResolvers by managedClusterSet
+		serviceResolvers = filterManagedProxyServiceResolverWithManagedClusterSet(serviceResolvers, managedClusterSetMap)
+
+		// convert serviceResolvers to servicesToExpose
+		servicesToExpose := []map[string]string{}
+		for _, serviceResolver := range serviceResolvers {
+			servicesToExpose = append(servicesToExpose, convertManagedProxyServiceResolverToService(cluster.Name, serviceResolver))
+		}
+		// remove dup and sort servicesToExpose
+		servicesToExpose = removeDupAndSortServices(servicesToExpose)
```

Comment:
could we use one loop to handle this and use set instead of []map[string]string{}?

```
servicesToExpose := k8s.io/apimachinery/pkg/util/sets.String{}
for _, serviceResolver := range serviceResolverList.Items {
		if !serviceResolver.DeletionTimestamp.IsZero() {
			continue
		}
                if !util.IsServiceResolverLegal(&serviceResolver) {
			continue
		}
               if _, ok := managedClusterSetMap[serviceResolver.Spec.ManagedClusterSelector.ManagedClusterSet.Name]; !ok {
			continue
		}

                servicesToExpose.Insert(util.GenerateServiceURL(clusterName, serviceResolver.Spec.ServiceSelector.ServiceRef.Namespace, serviceResolver.Spec.ServiceSelector.ServiceRef.Name))
} 

// add servicesToExpose into aids
for _, s := range servicesToExpose.List() {
     aids = append(aids, fmt.Sprintf("host=%s", s["url"]))
}

```

Reply:
Also, like the above mentioned, use different functions to make unit-test cases clearer.

And we need a **sorted** services list here, and `sets.String` can't promise this.

Reply:
I think  the `aids ` and `servicesToExpose` is sorted here

why we need to sort the `servicesToExpose`? for test? 

the problem of `sets.String` is we cannot render the `Service` helm chart easily, right?
and I thought this agin, could we use `map[string]ManagedProxyServiceResolver` to render the `Service` helm chart  instead of map[string]string?

```
{{- range $key, $val := .Values.servicesToExpose }}
apiVersion: v1
kind: Service
metadata:
  namespace: {{ $.Release.Namespace }}
  name: {{ $key }}
spec:
  type: ExternalName
  externalName: {{ $val.Spec.ServiceSelector.ServiceRef.Name}}.{{ $val.Spec.ServiceSelector.ServiceRef.Namespace}}.svc.cluster.local
---
{{- end }}
```

or consider the `ServiceSelector` may has other type, we may make the map[string]string more simple
map[string]string{
    "url": "",
   "externalName": ""
}

Reply:
The render part won't be an issue for sure, the case needs sort is when (s1, s2, s3) turns to (s2, s3, s1), we don't want agent to re-deploy the manifests.

Reply:
Update.